app/merginapi.cpp

```diff
@@ -42,18 +114,122 @@ void MerginApi::listProjectsReplyFinished()
     if (r->error() == QNetworkReply::NoError)
     {
       QByteArray data = r->readAll();
+      cacheProjectsData(data);
       mMerginProjects = parseProjectsData(data);
     }
     else {
         QString message = QStringLiteral("Network API error: %1(): %2").arg("listProjects", r->errorString());
         qDebug("%s", message.toStdString().c_str());
-        emit networkErrorOccurred( r->errorString() );
+        emit networkErrorOccurred( r->errorString(), "Mergin API error: listProjects" );
     }
 
     r->deleteLater();
     emit listProjectsFinished();
 }
 
+void MerginApi::downloadProjectReplyFinished()
+{
+
+    QNetworkReply* r = qobject_cast<QNetworkReply*>(sender());
+    if (!r) {
+        emit networkErrorOccurred( "No reply received.", "Mergin API error: downloadProject" );
+        qDebug() << r->errorString();
+        mPendingRequests.remove(r->url());
+        return;
+    }
+
+    if (r->error() == QNetworkReply::NoError)
+    {
+        QString projectName("temp");
+        if (mPendingRequests.contains(r->url())) {
+            projectName = mPendingRequests.value(r->url());
+        }
+        QString projectDir = mDataDir + projectName;
+        QFile file(projectDir);
+
+        handleDataStream(r, projectDir);
+        emit downloadProjectFinished(projectDir, projectName);
+        emit notify("Download successful");
+    }
+    else {
+        qDebug() << r->errorString();
+        emit networkErrorOccurred( r->errorString(), "Mergin API error: downloadProject" );
+    }
+    mPendingRequests.remove(r->url());
+    r->deleteLater();
+}
+
+void MerginApi::projectInfoReplyFinished()
+{
+    QNetworkReply* r = qobject_cast<QNetworkReply*>(sender());
+    if (!r) {
+        emit networkErrorOccurred( "No reply received.", "Mergin API error: projectInfo" );
+        return;
+    }
+
+    QList<MerginFile> files;
+    QUrl url = r->url();
+    QStringList res = url.path().split("/");
+    QString projectName;
+    projectName = res.last();
+    QString projectPath = QString(mDataDir + projectName + "/");
+
+    if (r->error() == QNetworkReply::NoError)
+    {
+      QByteArray data = r->readAll();
+      QJsonDocument doc = QJsonDocument::fromJson(data);
+      if (doc.isObject()) {
+          QJsonObject docObj = doc.object();
+          auto it = docObj.constFind("files");
+          QJsonValue v = *it;
+          Q_ASSERT( v.isArray() );
+          QJsonArray vArray = v.toArray();
+
+          QSet<QString> localFiles = listFiles(projectPath);
+          for ( auto it = vArray.constBegin(); it != vArray.constEnd(); ++it)
+          {
+            QJsonObject projectInfoMap = it->toObject();
+            QString serverChecksum = projectInfoMap.value("checksum").toString();
+            QString path = projectInfoMap.value("path").toString();
+            QString localChecksum = QString::fromStdString(getChecksum(projectPath + path).toStdString());
```

Comment:
no need to go through std::string - you can use e.g. QString::fromLatin1() to convert from QByteArray to QString